typed-racket-lib/typed-racket/base-env/prims-contract.rkt

```diff
@@ -250,9 +251,17 @@
 ;;  make-predicate
 ;;  cast
 
-;; Helper to construct syntax for contract definitions
+;; Helpers to construct syntax for contract definitions
+;; make-contract-def-rhs : Type-Stx Boolean Boolean -> Syntax
 (define (make-contract-def-rhs type flat? maker?)
-  (contract-def-property #'#f `#s(contract-def ,type ,flat? ,maker? untyped)))
+  (define contract-def `#s(contract-def ,type ,flat? ,maker? untyped))
+  (contract-def-property #'#f (λ () contract-def)))
```

Comment:
Just a quick overview of how this works:

* When you report a type error and we're reporting multiple type errors, it gets stored in a list in a mutable variable.
* Then, after `pass2` in the type checker, we call `report-all-errors`, which will raise a syntax error if needed. That code is [here](https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt#L409). 
* The contract generation happens after this, when we (generate the provides)[https://github.com/racket/typed-racket/blob/master/typed-racket-lib/typed-racket/typecheck/tc-toplevel.rkt#L445]. 

I'm ok with sticking with the `lambda`, but I'd like to make this an error if we call the thunk to early.